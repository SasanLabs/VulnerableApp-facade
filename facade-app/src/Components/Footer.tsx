import React from "react";
import {
  Footer as RSuiteFooter,
  FooterProps,
  Navbar as RSuiteNavBar,
} from "rsuite";

interface IFooter extends FooterProps {
  copyRightYear: number;
}

export default class Footer extends React.Component<IFooter> {
  render() {
    const { copyRightYear } = this.props;
    return (
      <RSuiteFooter>
        <RSuiteNavBar>
          <RSuiteNavBar.Body
            style={{ textAlign: "center", height: "30px", fontSize: "15px" }}
          >
            <div data-testid={"FOOTER_COPYRIGHT_TEXT"}>
              &copy; Copyright {copyRightYear}, SasanLabs
            </div>
          </RSuiteNavBar.Body>
        </RSuiteNavBar>
      </RSuiteFooter>
    );
  }
}
