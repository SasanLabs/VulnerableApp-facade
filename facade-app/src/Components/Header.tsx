import React from "react";
import VulnerableAppLogo from "../images/Logo.png";

import {
  Header as RSuiteHeader,
  Navbar as RSuiteNavBar,
  Nav as RSuiteNav,
  Icon as RSuiteIcon,
  Dropdown as RSuiteDropDown,
} from "rsuite";
import { Props } from "../interface/Props";

import "../styles/Header.css";

export default class Header extends React.Component<Props, {}> {
  render() {
    const { globalState, setGlobalState } = this.props;
    return (
      <RSuiteHeader>
        <RSuiteNavBar appearance="inverse">
          <RSuiteNavBar.Header>
            <img
              src={VulnerableAppLogo}
              width="55"
              height="55"
              alt="vulnerable app logo"
            />
          </RSuiteNavBar.Header>
          <RSuiteNavBar.Body>
            <RSuiteNav>
              <RSuiteNav.Item
                onSelect={() =>
                  setGlobalState({
                    ...globalState,
                    activateHomePage: true,
                    activateAboutUsPage: false,
                  })
                }
              >
                <b>Owasp VulnerableApp-Facade</b>
              </RSuiteNav.Item>
            </RSuiteNav>
            <RSuiteNav pullRight>
              <RSuiteNav.Item
                onSelect={() =>
                  setGlobalState({
                    ...globalState,
                    activateHomePage: true,
                    activateAboutUsPage: false,
                  })
                }
                icon={<RSuiteIcon icon="home" role={"img"} />}
              >
                Home
              </RSuiteNav.Item>
              <RSuiteNav.Item
                onSelect={() =>
                  setGlobalState({
                    ...globalState,
                    activateHomePage: false,
                    activateAboutUsPage: true,
                  })
                }
              >
                About Us
              </RSuiteNav.Item>
              <RSuiteDropDown title="Scanners">
                <RSuiteDropDown.Item title="Dynamic Application Security Testing">
                  DAST
                </RSuiteDropDown.Item>
                <RSuiteDropDown.Item title="Static Application Security Testing">
                  SAST
                </RSuiteDropDown.Item>
              </RSuiteDropDown>
              <a href="https://github.com/SasanLabs/VulnerableApp-facade">
                <RSuiteNav.Item
                  icon={<RSuiteIcon icon="github" role={"img"} />}
                >
                  Github
                </RSuiteNav.Item>
              </a>
              <RSuiteDropDown title="Projects by SasanLabs">
                <a href="https://github.com/SasanLabs/VulnerableApp">
                  <RSuiteDropDown.Item
                    icon={<RSuiteIcon icon="github" role={"img"} />}
                  >
                    Owasp VulnerableApp
                  </RSuiteDropDown.Item>
                </a>
                <a href="https://github.com/SasanLabs/owasp-zap-jwt-addon">
                  {" "}
                  <RSuiteDropDown.Item
                    icon={<RSuiteIcon icon="github" role={"img"} />}
                  >
                    ZAP JWT Addon
                  </RSuiteDropDown.Item>
                </a>
                <a href="https://github.com/SasanLabs/owasp-zap-fileupload-addon">
                  <RSuiteDropDown.Item
                    icon={<RSuiteIcon icon="github" role={"img"} />}
                  >
                    ZAP FileUpload Addon
                  </RSuiteDropDown.Item>
                </a>
              </RSuiteDropDown>
            </RSuiteNav>
          </RSuiteNavBar.Body>
        </RSuiteNavBar>
      </RSuiteHeader>
    );
  }
}
