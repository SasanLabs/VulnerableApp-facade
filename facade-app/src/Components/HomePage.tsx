import React from "react";
import { Panel as RSuitePanel, PanelGroup as RSuitePanelGroup } from "rsuite";

export class HomePage extends React.Component {
  Card = () => (
    <div data-testid={"HOME_PAGE_MAIN_CONTAINER"}>
      <RSuitePanel
        header="Welcome to Owasp VulnerableApp-Facade"
        className="VulnerableApp-Facade-HomePage-Content-Header"
      >
        <div
          className="VulnerableApp-Facade-HomePage-Content"
          data-testid={"HOME_PAGE_CONTENT_CONTAINER"}
        >
          As we are seeing a lot of technological enhancements in the industry
          from past few years, these technical enhancements are solving one or
          the other problem however, with that they also bring few different
          vulnerabilities. Vulnerable Applications are generally written in one
          of the techstacks like either Node.js or Java with a SQL or NoSQL
          database etc and hence they are not able to expand to a whole new set
          of vulnerabilities which are present in other technologies. Also
          adding more vulnerabilities in a single vulnerable application makes
          it heavier and complex which finally makes it unmaintainable. So
          VulnerableApp-facade is built to solve this problem by building a
          distributed farm of Vulnerable Applications such that they can be
          built agnostic to tech stacks.
          <p>Following is the design diagram of Owasp VulnerableApp-Facade:</p>
          <img
            width="80%"
            src="/images/VulnerableApp-facade.jpeg"
            alt="Unable to lead the design diagram"
          />
          <p>
            Here VulnerableApp-Facade is running as a gateway or a proxy which
            is routing calls to actual Vulnerable Applications based on a
            criteria defined in the nginx configuration.
          </p>
        </div>
      </RSuitePanel>
      <RSuitePanel
        className="VulnerableApp-Facade-HomePage-Content-Header"
        header="Warning"
        data-testid={"HOME_PAGE_WARNING_CONTAINER"}
      >
        <div className="VulnerableApp-Facade-HomePage-Content">
          As VulnerableApp-Facade is a proxy wrapper over the actual Vulnerable
          Applications which can be very dangerous if exposed over the public
          internet. So we suggest you to please run it in local dev environments
          or environments without any public internet access.
        </div>
      </RSuitePanel>
    </div>
  );

  render() {
    return (
      <div>
        <RSuitePanelGroup>
          <this.Card />
        </RSuitePanelGroup>
      </div>
    );
  }
}
