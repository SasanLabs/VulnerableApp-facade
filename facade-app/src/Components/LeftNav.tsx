import React from "react";
import {
  Sidenav as RSuiteSidenav,
  Dropdown as RSuiteDropdown,
  Nav as RSuiteNav,
  Icon as RSuiteIcon,
} from "rsuite";
import { Props } from "../interface/Props";
import { LevelInformation, VulnerabilityDefinition } from "../interface/State";

export default class LeftNav extends React.Component<Props, {}> {
  constructor(props: Props) {
    super(props);
    this._handleVulnerabilityLevelSelect =
      this._handleVulnerabilityLevelSelect.bind(this);
    this._handleVulnerabilityTypeSelect =
      this._handleVulnerabilityTypeSelect.bind(this);
  }

  _handleVulnerabilityLevelSelect(
    applicationName: string,
    vulnerabilityName: string,
    levelIdentifier: string
  ) {
    this.props.setGlobalState({
      activeApplication: applicationName,
      activeVulnerability: vulnerabilityName,
      activeLevel: levelIdentifier,
      activateHomePage: false,
      activateAboutUsPage: false,
      showHints: false,
    });
  }

  _handleVulnerabilityTypeSelect(
    applicationName: string,
    vulnerabilityName: string
  ) {
    this.props.setGlobalState({
      activeApplication: applicationName,
      activeVulnerability: vulnerabilityName,
      activeLevel: undefined,
      activateHomePage: false,
      activateAboutUsPage: false,
      showHints: false,
    });
  }

  _getApplicationVulnerabilityLevels(
    applicationName: string,
    vulnerabilityName: string,
    levels: Array<LevelInformation>
  ) {
    const { activeApplication, activeVulnerability, activeLevel } =
      this.props.globalState;
    return (
      <RSuiteDropdown
        eventKey={applicationName + "." + vulnerabilityName}
        title={vulnerabilityName}
        activeKey={
          activeApplication + "." + activeVulnerability + "." + activeLevel
        }
        onOpen={() =>
          this._handleVulnerabilityTypeSelect(
            applicationName,
            vulnerabilityName
          )
        }
        className="VulnerableApp-Facade-LeftNav-Vulnerability-Level"
        data-testid={applicationName + "." + vulnerabilityName}
      >
        {levels.map((vulnerabilityLevel) => (
          <RSuiteDropdown.Item
            eventKey={
              applicationName +
              "." +
              vulnerabilityName +
              "." +
              vulnerabilityLevel.levelIdentifier
            }
            icon={
              vulnerabilityLevel.variant === "SECURE" ? (
                <RSuiteIcon icon="lock" />
              ) : (
                <RSuiteIcon icon="unlock" />
              )
            }
            onSelect={() =>
              this._handleVulnerabilityLevelSelect(
                applicationName,
                vulnerabilityName,
                vulnerabilityLevel.levelIdentifier
              )
            }
            data-testid={
              applicationName +
              "." +
              vulnerabilityName +
              "." +
              vulnerabilityLevel.levelIdentifier
            }
          >
            {vulnerabilityLevel.levelIdentifier}
          </RSuiteDropdown.Item>
        ))}
      </RSuiteDropdown>
    );
  }

  _getApplicationVulnerabilities(
    applicationName: string,
    vulnerabilityDefinition: Array<VulnerabilityDefinition>,
    activeApplication?: string
  ) {
    return (
      <RSuiteSidenav
        defaultOpenKeys={activeApplication ? [activeApplication] : []}
      >
        <RSuiteSidenav.Body>
          <RSuiteNav>
            <RSuiteDropdown
              eventKey={applicationName}
              title={applicationName}
              className="VulnerableApp-Facade-LeftNav-Application"
              icon={<RSuiteIcon icon="server" role={"img"} />}
            >
              <RSuiteDropdown.Item divider />
              {vulnerabilityDefinition.map((vulnerabilityDefinition) =>
                this._getApplicationVulnerabilityLevels(
                  applicationName,
                  vulnerabilityDefinition.name,
                  vulnerabilityDefinition.levels
                )
              )}
            </RSuiteDropdown>
          </RSuiteNav>
        </RSuiteSidenav.Body>
      </RSuiteSidenav>
    );
  }

  render() {
    const { isSuccessfullyLoaded, applicationData, activeApplication } =
      this.props.globalState;
    if (isSuccessfullyLoaded && applicationData) {
      return (
        <div data-testid={"LEFT_NAV_CONTAINER"}>
          {applicationData.map((applicationData) => {
            let applicationName = applicationData.applicationName;
            let vulnerabilityDefinitions =
              applicationData.vulnerabilityDefinitions;
            return this._getApplicationVulnerabilities(
              applicationName,
              vulnerabilityDefinitions,
              activeApplication
            );
          })}
        </div>
      );
    }
    return <div />;
  }
}
