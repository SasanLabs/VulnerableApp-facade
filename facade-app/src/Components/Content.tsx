import React from "react";
import { LevelInformation } from "../interface/State";
import { Panel as RSuitePanel } from "rsuite";
import {
  appendStaticResourcesToDocument,
  getResource,
  manipulateDOM,
} from "../Utilities/Utils";
import { VulnerabilityDefinitionResponse } from "../interface/GeneralContracts";
import { HomePage } from "./HomePage";
import AboutUs from "./AboutUs";
import { Props } from "../interface/Props";

export class Content extends React.Component<Props> {
  selectedLevel?: LevelInformation;

  componentDidUpdate(prevProps: Props) {
    if (
      prevProps.globalState.activeLevel !==
        this.props.globalState.activeLevel ||
      prevProps.globalState.activeVulnerability !==
        this.props.globalState.activeVulnerability ||
      prevProps.globalState.activeApplication !==
        this.props.globalState.activeApplication
    ) {
      const {
        activeApplication,
        applicationData,
        activeVulnerability,
        activeLevel,
      } = this.props.globalState;
      const selectedApplicationState = applicationData?.find(
        (applicationState) =>
          applicationState.applicationName === activeApplication
      );
      if (selectedApplicationState) {
        const selectedVulnerability =
          selectedApplicationState.vulnerabilityDefinitions.find(
            (vulnerabilityDefinition) =>
              vulnerabilityDefinition.id === activeVulnerability
          );
        this.setState({ description: selectedVulnerability?.description });
        if (selectedVulnerability) {
          manipulateDOM(
            "__vuln_description__",
            selectedVulnerability.description
          );
          const selectedLevel = selectedVulnerability.levels.find(
            (level) => level.levelIdentifier === activeLevel
          );
          if (selectedLevel) {
            this.selectedLevel = selectedLevel;
            getResource(
              selectedLevel.resourceInformation.htmlResource.uri,
              this._setLocalState.bind(this),
              false
            );
          }
        }
      }
    }
  }

  _setLocalState(
    vulnerabilityDefinitionResponse: VulnerabilityDefinitionResponse
  ) {
    if (vulnerabilityDefinitionResponse.data) {
      manipulateDOM("__content__", vulnerabilityDefinitionResponse.data);
      if (this.selectedLevel) {
        appendStaticResourcesToDocument(this.selectedLevel);
      }
    }
  }

  render() {
    const {
      activeVulnerability,
      activeApplication,
      activeLevel,
      activateHomePage,
      activateAboutUsPage,
      showHints,
    } = this.props.globalState;
    const { setGlobalState } = this.props;
    return (
      <div className="VulnerableApp-Facade-Info">
        {activateHomePage ? (
          <HomePage></HomePage>
        ) : activateAboutUsPage ? (
          <AboutUs></AboutUs>
        ) : (
          <div className="VulnerableApp-Facade-Content-Container">
            {activeVulnerability ? (
              <div>
                <RSuitePanel
                  header="Vulnerability Description"
                  className="VulnerableApp-Facade-Content-Vulnerability-Description-Header"
                  collapsible={true}
                  defaultExpanded={true}
                >
                  <div className="VulnerableApp-Facade-Content">
                    <div
                      id="__vuln_description__"
                      data-testid={"VULNERABILITY_CONTENT_DESCRIPTION"}
                    />
                  </div>
                </RSuitePanel>
              </div>
            ) : (
              <div />
            )}
            {activeVulnerability ? (
              <div>
                <RSuitePanel
                  header="Practice Vulnerability"
                  className="VulnerableApp-Facade-Content-Practice-Vulnerability-Header"
                >
                  <div className="VulnerableApp-Facade-Content">
                    <div
                      id="__content__"
                      data-testid={"VULNERABILITY_MAIN_CONTENT"}
                    />
                  </div>
                </RSuitePanel>
                {this.selectedLevel &&
                this.selectedLevel.hints &&
                this.selectedLevel.hints.length > 0 ? (
                  <RSuitePanel
                    header="Hints"
                    className="VulnerableApp-Facade-Content-Hint-Content"
                    collapsible={true}
                    defaultExpanded={false}
                    expanded={showHints}
                    onSelect={() =>
                      setGlobalState({
                        activeApplication: activeApplication,
                        activeVulnerability: activeVulnerability,
                        activeLevel: activeLevel,
                        activateHomePage: false,
                        activateAboutUsPage: false,
                        showHints: !showHints,
                      })
                    }
                  >
                    <ol data-testid={"VULNERABILITY_HINTS"}>
                      {this.selectedLevel.hints.map((hint) => {
                        return <li>{hint.description}</li>;
                      })}
                    </ol>
                  </RSuitePanel>
                ) : (
                  <div />
                )}
              </div>
            ) : (
              <div></div>
            )}
          </div>
        )}
      </div>
    );
  }
}
