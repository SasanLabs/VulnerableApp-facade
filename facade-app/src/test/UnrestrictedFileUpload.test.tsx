import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { Content } from "../Components/Content";
import testFixture from "./fixtures";

jest.mock("../Utilities/Utils", () => {
  const actual = jest.requireActual("../Utilities/Utils");
  return {
    ...actual,
    getResource: jest.fn(),
  };
});

describe("UnrestrictedFileUpload content", () => {
  const vulnerabilityLevels = [
    "LEVEL_1",
    "LEVEL_2",
    "LEVEL_3",
    "LEVEL_4",
    "LEVEL_5",
    "LEVEL_6",
  ];

  it("should render correctly", () => {
    const mock = () => jest.fn();
    const state = {
      ...testFixture,
      activeApplication: "VulnerableApp",
      activeVulnerability: "UnrestrictedFileUpload",
      activeLevel: "LEVEL_1",
    };
    const { container } = render(
      // @ts-ignore
      <Content globalState={state} setGlobalState={mock} />
    );

    expect(container).toMatchSnapshot();
  });

  it.each(vulnerabilityLevels)(
    `%s should render vulnerability description`,
    async (level) => {
      const mock = () => jest.fn();
      const state = {
        ...testFixture,
        activeApplication: "VulnerableApp",
        activeVulnerability: "UnrestrictedFileUpload",
        activeLevel: level,
      };
      const { rerender } = render(
        // @ts-ignore
        <Content globalState={testFixture} setGlobalState={mock} />
      );

      // @ts-ignore
      rerender(<Content globalState={state} setGlobalState={mock} />);

      const header = screen.getByText("Vulnerability Description");
      expect(header).toBeInTheDocument();

      await waitFor(() => {
        const description = screen.getByTestId(
          "VULNERABILITY_CONTENT_DESCRIPTION"
        );
        const link1 = screen.getByText("Owasp Wiki Link");
        const link2 = screen.getByText(
          "Ebrahim Hegazy talk on Unrestricted File Uploads"
        );
        const link3 = screen.getByText(
          "Sans rules to implement secure file uploads"
        );

        expect(description).toHaveTextContent(
          "Uploaded files represent a significant risk to applications. The first step in many attacks is to get some code to the system to be attacked. Then the attack only needs to find a way to get the code executed.The consequences of unrestricted file upload can vary, including complete system takeover, an overloaded file system or database, forwarding attacks to back-end systems, client-side attacks, or simple defacement. It depends on what the application does with the uploaded file and especially where it is stored.Important Links: Owasp Wiki Link Ebrahim Hegazy talk on Unrestricted File Uploads Sans rules to implement secure file uploads"
        );
        expect(link1).toBeInTheDocument();
        expect(link1).toHaveAttribute(
          "href",
          "https://owasp.org/www-community/vulnerabilities/Unrestricted_File_Upload"
        );
        expect(link2).toBeInTheDocument();
        expect(link2).toHaveAttribute(
          "href",
          "https://www.youtube.com/watch?v=CmF9sEyKZNo"
        );
        expect(link3).toBeInTheDocument();
        expect(link3).toHaveAttribute(
          "href",
          "https://www.sans.org/blog/8-basic-rules-to-implement-secure-file-uploads/"
        );
      });
    }
  );

  it.each(vulnerabilityLevels)(
    `%s should render practice vulnerability section`,
    async (level) => {
      const mock = () => jest.fn();
      const state = {
        ...testFixture,
        activeApplication: "VulnerableApp",
        activeVulnerability: "UnrestrictedFileUpload",
        activeLevel: level,
      };
      const { rerender } = render(
        // @ts-ignore
        <Content globalState={testFixture} setGlobalState={mock} />
      );

      // @ts-ignore
      rerender(<Content globalState={state} setGlobalState={mock} />);

      await waitFor(() => {
        const header = screen.getByText("Practice Vulnerability");
        expect(header).toBeInTheDocument();
      });
    }
  );

  const input = [
    ["LEVEL_1", "There is no validation on uploaded file's name."],
    ["LEVEL_2", "There is no validation on uploaded file's name."],
    ["LEVEL_3", "All file extensions are allowed except .html extensions."],
    [
      "LEVEL_4",
      "All file extensions are allowed except .html and .htm extensions.",
    ],
    [
      "LEVEL_5",
      "All file extensions are allowed except case insensitive .html, .htm extensions.",
    ],
    [
      "LEVEL_5",
      "All file extensions are allowed except case insensitive .html, .htm extensions.",
    ],
    [
      "LEVEL_6",
      "Only file name is allowed if it contains case insensitive .jpeg or .png.",
    ],
    [
      "LEVEL_7",
      "Only file name is allowed if it ends with case insensitive .jpeg or .png and it is considered before Null Bytes only.",
    ],
    ["LEVEL_8", "There is no validation on uploaded file's name."],
    [
      "LEVEL_9",
      "Only file name is allowed if it ends with case insensitive .jpeg or .png.",
    ],
  ];

  it.each(input)(`%s should render hints`, async (level, hint) => {
    const mock = () => jest.fn();
    const state = {
      ...testFixture,
      activeApplication: "VulnerableApp",
      activeVulnerability: "UnrestrictedFileUpload",
      activeLevel: level,
      showHints: true,
    };
    const { rerender } = render(
      // @ts-ignore
      <Content globalState={testFixture} setGlobalState={mock} />
    );

    // @ts-ignore
    rerender(<Content globalState={state} setGlobalState={mock} />);

    await waitFor(() => {
      const header = screen.getByText("Hints");
      expect(header).toBeInTheDocument();

      const hintText = screen.getByText(hint);
      expect(hintText).toBeInTheDocument();
    });
  });
});
