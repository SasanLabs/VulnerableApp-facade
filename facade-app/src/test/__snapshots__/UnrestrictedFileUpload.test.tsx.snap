// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnrestrictedFileUpload content should render correctly 1`] = `
<div>
  <div
    class="VulnerableApp-Facade-Info"
  >
    <div
      class="VulnerableApp-Facade-Content-Container"
    >
      <div>
        <div
          class="VulnerableApp-Facade-Content-Vulnerability-Description-Header rs-panel rs-panel-default rs-panel-in rs-panel-collapsible"
        >
          <div
            class="rs-panel-heading"
            role="rowheader"
            tabindex="-1"
          >
            <span
              class="rs-panel-title"
              role="presentation"
            >
              <span
                aria-controls="undefined"
                aria-expanded="true"
                aria-selected="true"
              >
                Vulnerability Description
              </span>
            </span>
          </div>
          <div
            aria-hidden="false"
            class="rs-panel-collapse VulnerableApp-Facade-Content-Vulnerability-Description-Header collapse in"
          >
            <div
              class="rs-panel-body"
            >
              <div
                class="VulnerableApp-Facade-Content"
              >
                <div
                  data-testid="VULNERABILITY_CONTENT_DESCRIPTION"
                  id="__vuln_description__"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="VulnerableApp-Facade-Content-Practice-Vulnerability-Header rs-panel rs-panel-default"
        >
          <div
            class="rs-panel-heading"
            role="rowheader"
            tabindex="-1"
          >
            Practice Vulnerability
          </div>
          <div
            class="rs-panel-body"
          >
            <div
              class="VulnerableApp-Facade-Content"
            >
              <div
                data-testid="VULNERABILITY_MAIN_CONTENT"
                id="__content__"
              />
            </div>
          </div>
        </div>
        <div />
      </div>
    </div>
  </div>
</div>
`;
